Sources/SwiftFusion/Core/MatrixN.swift

```diff
@@ -1,8 +1,106 @@
-public struct Matrix3: Differentiable & Equatable & KeyPathIterable {
+public struct Matrix3: Differentiable & Equatable & KeyPathIterable & EuclideanVectorN & AdditiveArithmetic {
```

Comment:
Wow, talk about boilerplate :-) I am OK with merging this for this PR but having matrices and/or tensors be automatically differentiable seems something that should come with "swift for tensorflow". I know we're talking here about small matrices, which we might have to do differently - but the amount of code here seems like a large cost - and that's only Matrix3 ! @marcrasi @dabrahams take a look, have a think :-)

Reply:
Yeah, we can definitely remove a lot of boilerplate by replacing this Matrix3 with a generic FixedSizeMatrix type that has differentiability and other matrix operations on it. It's really useful to have this hardcoded boilerplaty thing checked in for now, so that we can see what features are important to add first.

Reply:
Yup, let's merge this first :)